src/libraries/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs

```diff
@@ -102,8 +131,36 @@ private static int SignalAndWaitCore(IntPtr handleToSignal, IntPtr handleToWaitO
         {
             Debug.Assert(millisecondsTimeout >= -1);
 
-            int ret = (int)Interop.Kernel32.SignalObjectAndWait(handleToSignal, handleToWaitOn, (uint)millisecondsTimeout, Interop.BOOL.FALSE);
+            ulong startTime = 0;
+            if (millisecondsTimeout != -1)
+            {
+                startTime = Interop.Kernel32.GetTickCount64();
+            }
+
+        retry:
+            int ret = (int)Interop.Kernel32.SignalObjectAndWait(handleToSignal, handleToWaitOn, (uint)millisecondsTimeout, Interop.BOOL.TRUE);
+
+            if (ret == Interop.Kernel32.WAIT_IO_COMPLETION)
+            {
+                // Handle APC completion by adjusting timeout and retrying
+                if (millisecondsTimeout != -1)
+                {
+                    ulong currentTime = Interop.Kernel32.GetTickCount64();
+                    ulong elapsed = currentTime - startTime;
+                    
+                    if (elapsed >= (ulong)millisecondsTimeout)
+                    {
+                        ret = Interop.Kernel32.WAIT_TIMEOUT;
+                        goto WaitCompleted;
+                    }
+                    
+                    millisecondsTimeout -= (int)elapsed;
+                    startTime = currentTime;
+                }
+                goto retry;
+            }
 
+        WaitCompleted:
```

Comment:
Could you please refactor the code by switching the if condition above to make WaitCompleted label unnecessary?